provider/src/main/kotlin/org/gradle/kotlin/dsl/codegen/ApiTypeProvider.kt

```diff
@@ -0,0 +1,448 @@
+
+    private
+    var currentTypeParameter: String? = null
+
+    override fun visitFormalTypeParameter(binaryName: String) {
+        typeParameters[binaryName] = ArrayList(1)
+        currentTypeParameter = binaryName
+    }
+
+    override fun visitClassBound(): SignatureVisitor =
+        TypeSignatureVisitor().also { typeParameters[currentTypeParameter]!!.add(it) }
+
+    override fun visitInterfaceBound(): SignatureVisitor =
+        TypeSignatureVisitor().also { typeParameters[currentTypeParameter]!!.add(it) }
+}
+
+
+private
+class ClassSignatureVisitor : BaseSignatureVisitor()
+
+
+private
+class MethodSignatureVisitor : BaseSignatureVisitor() {
+
+    val parameters: MutableList<TypeSignatureVisitor> = ArrayList(1)
+
+    val returnType = TypeSignatureVisitor()
+
+    override fun visitParameterType(): SignatureVisitor =
+        TypeSignatureVisitor().also { parameters.add(it) }
+
+    override fun visitReturnType(): SignatureVisitor =
+        returnType
+}
+
+
+private
+class TypeSignatureVisitor : SignatureVisitor(ASM6) {
+
+    var isArray = false
+
+    lateinit var binaryName: String
+
+    val typeArguments = ArrayList<TypeSignatureVisitor>(1)
+
+    private
+    var expectTypeArgument = false
+
+    override fun visitBaseType(descriptor: Char) =
+        visitBinaryName(binaryNameOfBaseType(descriptor))
+
+    override fun visitArrayType(): SignatureVisitor =
+        TypeSignatureVisitor().also {
+            visitBinaryName("kotlin.Array")
+            isArray = true
+            typeArguments.add(it)
+        }
+
+    override fun visitClassType(internalName: String) =
+        visitBinaryName(binaryNameOfInternalName(internalName))
+
+    override fun visitInnerClassType(localName: String) {
+        binaryName += "${'$'}$localName"
+    }
+
+    override fun visitTypeArgument() {
+        typeArguments.add(TypeSignatureVisitor().also { it.binaryName = "?" })
+    }
+
+    override fun visitTypeArgument(wildcard: Char): SignatureVisitor =
+        TypeSignatureVisitor().also {
+            expectTypeArgument = true
+            typeArguments.add(it)
+        }
+
+    override fun visitTypeVariable(internalName: String) {
+        visitBinaryName(binaryNameOfInternalName(internalName))
+    }
+
+    private
+    fun visitBinaryName(binaryName: String) {
+        if (expectTypeArgument) {
+            TypeSignatureVisitor().let {
+                typeArguments.add(it)
+                SignatureReader(binaryName).accept(it)
+            }
+            expectTypeArgument = false
+        } else {
+            this.binaryName = binaryName
+        }
+    }
+}
+
+
+private
+fun binaryNameOfBaseType(descriptor: Char) =
+    Type.getType(descriptor.toString()).className
+
+
+private
+fun binaryNameOfInternalName(internalName: String): String =
+    Type.getObjectType(internalName).className
+
+
+private
+fun sourceNameOfBinaryName(binaryName: String): String =
```

Comment:
Is this function taking method names, class names, or fully qualified method names?

Reply:
It is taking binary names of classes, see https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html#jls-13.1

Reply:
Any need to worry about anonymous inner classes generated by the compiler with names like:
`testing/MinimalRequiredCodeTest$simpleExampleCase$1$1`?

Just thinking about a bug I faced with the Junit5 Classloader that had a bug:
https://github.com/gradle/gradle/issues/4921

Reply:
Good point, I added filtering and test coverage for this.